test/solver/batch_cg_kernels.cpp

```diff
@@ -0,0 +1,240 @@
+
+    auto iter_array = res.log_data->iter_counts.get_const_data();
+    for (size_t i = 0; i < num_batch_items; i++) {
+        ASSERT_EQ(iter_array[i], ref_iters);
+    }
+}
+
+
+TEST_F(BatchCg, CanSolve3ptStencilSystem)
+{
+    const int num_batch_items = 8;
+    const int num_rows = 100;
+    const int num_rhs = 1;
+    const real_type tol = 1e-5;
+    const int max_iters = 500;
+    auto mat =
+        gko::share(gko::test::generate_3pt_stencil_batch_matrix<const Mtx>(
+            exec, num_batch_items, num_rows));
+    auto linear_system = setup_linsys_and_solver(mat, num_rhs, tol, max_iters);
+    auto solver = gko::share(solver_factory->generate(linear_system.matrix));
+
+    auto res = gko::test::solve_linear_system(exec, linear_system, solver);
+
+    GKO_ASSERT_BATCH_MTX_NEAR(res.x, linear_system.exact_sol, tol * 50);
+    for (size_t i = 0; i < num_batch_items; i++) {
+        auto comp_res_norm = res.host_res_norm->get_const_values()[i] /
+                             linear_system.host_rhs_norm->get_const_values()[i];
+        ASSERT_LE(comp_res_norm, tol * 50);
+    }
+}
+
+
+TEST_F(BatchCg, CanSolveLargeBatchSizeHpdSystem)
+{
+    const int num_batch_items = 100;
+    const int num_rows = 102;
+    const int num_rhs = 1;
+    const real_type tol = 1e-5;
+    const int max_iters = num_rows * 2;
+    std::shared_ptr<Logger> logger = Logger::create();
+    auto mat =
+        gko::share(gko::test::generate_diag_dominant_batch_matrix<const Mtx>(
+            exec, num_batch_items, num_rows, true));
+    auto linear_system = setup_linsys_and_solver(mat, num_rhs, tol, max_iters);
+    auto solver = gko::share(solver_factory->generate(linear_system.matrix));
+    solver->add_logger(logger);
+
+    auto res = gko::test::solve_linear_system(exec, linear_system, solver);
+
+    solver->remove_logger(logger);
+    auto iter_counts = gko::make_temporary_clone(exec->get_master(),
+                                                 &logger->get_num_iterations());
+    auto res_norm = gko::make_temporary_clone(exec->get_master(),
+                                              &logger->get_residual_norm());
+    GKO_ASSERT_BATCH_MTX_NEAR(res.x, linear_system.exact_sol, tol * 50);
+    for (size_t i = 0; i < num_batch_items; i++) {
+        auto comp_res_norm = res.host_res_norm->get_const_values()[i] /
+                             linear_system.host_rhs_norm->get_const_values()[i];
+        ASSERT_LE(iter_counts->get_const_data()[i], max_iters);
+        EXPECT_LE(res_norm->get_const_data()[i] /
+                      linear_system.host_rhs_norm->get_const_values()[i],
+                  tol * 20);
+        EXPECT_GT(res_norm->get_const_data()[i], real_type{0.0});
+        ASSERT_LE(comp_res_norm, tol * 50);
+    }
+}
+
+
+TEST_F(BatchCg, CanSolveLargeMatrixSizeHpdSystem)
+{
+    const int num_batch_items = 12;
+    const int num_rows = 1025;
+    const int num_rhs = 1;
+    const real_type tol = 1e-5;
+    const int max_iters = num_rows * 2;
```

Comment:
num_rows should be the enough in theory, right?

Reply:
Yes, but not in practice. Especially without a preconditioner and for matrices with eigenvalues very close to each other.